src/i_pcsound.c

```diff
@@ -0,0 +1,522 @@
+    alSourcei(callback_source, AL_SOURCE_RELATIVE, AL_TRUE);
+    alSourcei(callback_source, AL_ROLLOFF_FACTOR, 0);
+    alSourcePlay(callback_source);
+    if (alGetError() != AL_NO_ERROR)
+    {
+        I_Printf(VB_ERROR, "RegisterCallback: Failed start playback.");
+        return;
+    }
+}
+
+static void UnregisterCallback(void)
+{
+    alSourceStop(callback_source);
+    alDeleteSources(1, &callback_source);
+    alDeleteBuffers(1, &callback_buffer);
+}
+
+static boolean I_PCS_ReinitSound(void)
+{
+    mixing_freq = SND_SAMPLERATE;
+
+    current_freq = 0;
+    current_remaining = 0;
+
+    RegisterCallback();
+
+    if (!sound_lock)
+    {
+        sound_lock = SDL_CreateMutex();
+    }
+
+    return true;
+}
+
+static boolean I_PCS_InitSound(void)
+{
+    if (!I_OAL_InitSound())
+    {
+        return false;
+    }
+
+    I_PCS_ReinitSound();
+
+    return true;
+}
+
+static void I_PCS_ShutdownModule(void)
+{
+    int i;
+
+    for (i = 0; i < num_sfx; ++i)
+    {
+        S_sfx[i].lumpnum = -1;
+    }
+
+    UnregisterCallback();
+}
+
+static void I_PCS_ShutdownSound(void)
+{
+    I_PCS_ShutdownModule();
+
+    I_OAL_ShutdownSound();
+}
+
+static boolean I_PCS_CacheSound(sfxinfo_t *sfx)
+{
+    if (IsDisabledSound(sfx))
+    {
+        return false;
+    }
+
+    return (GetLumpNum(sfx) != -1);
+}
+
+static boolean I_PCS_AdjustSoundParams(const mobj_t *listener, const mobj_t *source,
+                                      int chanvol, int *vol, int *sep, int *pri)
+{
+    fixed_t adx, ady, dist;
+
+    // haleyjd 05/29/06: allow per-channel volume scaling
+    *vol = (snd_SfxVolume * chanvol) / 15;
+
+    if (*vol < 1)
+        return false;
+    else if (*vol > 127)
+        *vol = 127;
+
+    if (!source || source == players[displayplayer].mo)
+        return true;
+
+    // haleyjd 08/12/04: we cannot adjust a sound for a NULL listener.
+    if (!listener)
+        return true;
+
+    // calculate the distance to sound origin
+    //  and clip it if necessary
+    //
+    // killough 11/98: scale coordinates down before calculations start
+    // killough 12/98: use exact distance formula instead of approximation
+
+    adx = abs((listener->x >> FRACBITS) - (source->x >> FRACBITS));
+    ady = abs((listener->y >> FRACBITS) - (source->y >> FRACBITS));
+
+    if (ady > adx)
+        dist = adx, adx = ady, ady = dist;
+
+    dist = adx ? FixedDiv(adx, finesine[(tantoangle[FixedDiv(ady, adx) >> DBITS]
+                                         + ANG90) >> ANGLETOFINESHIFT]) : 0;
+
+    if (!dist)  // killough 11/98: handle zero-distance as special case
+        return true;
+
+    if (dist > S_CLIPPING_DIST >> FRACBITS)
+        return false;
+
+    // volume calculation
+    if (dist > S_CLOSE_DIST >> FRACBITS)
+        *vol = *vol * ((S_CLIPPING_DIST >> FRACBITS) - dist) / S_ATTENUATOR;
+
+    // haleyjd 09/27/06: decrease priority with volume attenuation
+    *pri = *pri + (127 - *vol);
```

Comment:
I wouldn't mind if you took the vanilla pendant of that code (i.e. without MBF's priority calculation) from Choco or Crispy for the pc-speaker emulsion module.

Reply:
I tried the vanilla `S_AdjustSoundParams` from Crispy, it fixes the weird beep in the PL2 demo2. Hard to tell, but it seems to work better. https://github.com/fabiangreffrath/woof/pull/1200/commits/d6528b1385f6819c14fcef9adb588018aed97b1b

Reply:
There are some differences but it sounds okay to me. Left channel is Crispy, right channel is Woof:

https://github.com/fabiangreffrath/woof/assets/56656010/3de2309a-4507-420f-9faf-3db8ed58ad92

Crispy will sound slightly different depending on if the framerate is uncapped. In this example, both are at 140 fps, although Woof will always update sound at 35 fps.

Reply:
Do they sound the same with frame rate capped at 35 fps?

Reply:
> Do they sound the same with frame rate capped at 35 fps?

Crispy and Woof are still different from each other:

![crispy_woof_pcs](https://github.com/fabiangreffrath/woof/assets/56656010/10503641-a474-4d00-a5ff-73bab65252ba)

Reply:
Hm, strange. Could this be caused by `I_DeferUpdates()`?

Reply:
Cool visualization as always 👍 Honestly, I'm surprised the difference is so small. But, I think both Crispy and Woof should always update the sound at 35 fps. So should we fix this in Crispy?

Reply:
Yes, please. "Uncapped framerate" should really only apply to rendering framerate and not change game physics at all.

Reply:
> Could this be caused by I_DeferUpdates()?

I can't check right now but it would be as simple as commenting out the two lines in `S_UpdateSounds()` and comparing with Audacity or something similar. However, if these differences are due to buffer sizes, or thread timing for OpenAL Soft (Woof) vs. SDL (Crispy), I don't know if it's possible to do much better (other than fixing the obvious wipe issue).

Reply:
In Woof the buffer size is approximately `3528`, in Crispy `8192`. OpenAL is usually more accurate.

Crispy fix: https://github.com/fabiangreffrath/crispy-doom/pull/1099